### PR TITLE
feat(server): adapter layer for asset and file

### PR DESCRIPTION
**Pull request by nourbalaha**
In this PR I added the following:
1- Asset adapter layer.
2- Implementation of file-upload for FS and GCS.
3- Added unit tests for adapter layer.

### CODE REVIEW

File: server/internal/usecase/interactor/asset.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	return i.repos.Asset.FindByIDs(ctx, assets)
+}
+
+func (i *Asset) FindByProject(ctx context.Context, pid id.ProjectID, keyword *string, sort *asset.SortType, p *usecase.Pagination, operator *usecase.Operator) ([]*asset.Asset, *usecase.PageInfo, error) {
+	pp, err := i.repos.Project.FindByID(ctx, pid)
+	if err != nil {
+		return nil, nil, err
+	}
+	return Run2(
+		ctx, nil, i.repos,
```
Comment (yk-eukarya) [inline server/internal/usecase/interactor/asset.go:38]:
```suggestion
		ctx, operator, i.repos,
```

File: server/internal/infrastructure/mongo/mongodoc/asset.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	pid, err := id.ProjectIDFrom(d.Project)
+	if err != nil {
+		return nil, err
+	}
+	uid, err := id.UserIDFrom(d.CreatedBy)
+	if err != nil {
+		return nil, err
+	}
+
+	return asset.New().
```
Comment (yk-eukarya) [inline server/internal/infrastructure/mongo/mongodoc/asset.go:81]:
```suggestion
        f := asset.NewFile(). ....
	return asset.New().
	File(f).
```
You need to implement a file builder and use here to set the model file when its returned from the db

File: server/internal/infrastructure/gcs/file.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	if file.Size >= fileSizeLimit {
+		return nil, gateway.ErrFileTooLarge
+	}
+
+	sn := sanitize.Path(newAssetID() + path.Ext(file.Path))
+	if sn == "" {
+		return nil, gateway.ErrInvalidFile
+	}
+
+	filename := path.Join(gcsAssetBasePath, sn)
```
Comment (rot1024) [inline server/internal/infrastructure/gcs/file.go:76]: specification is not followed

`gs://{bucket name}/{first 2 chars of uuid}/{rest of uuid}/{filename with ext}`

File: server/internal/infrastructure/mongo/asset_test.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+// 	repo := NewAsset(client)
+// 	ctx := context.Background()
+// 	err := repo.Save(ctx, expected)
+// 	assert.NoError(t, err)
+
+// 	got, err := repo.FindByID(ctx, expected.ID())
+// 	assert.NoError(t, err)
+// 	assert.Equal(t, expected, got)
+// }
```
Comment (rot1024) [inline server/internal/infrastructure/mongo/asset_test.go:33]: Write tests

File: server/schemas/asset.graphql [hunk 5]
```diff
@@ -1,9 +1,9 @@
 type Asset implements Node {
   id: ID!
-  project: Project!
+  project: Project @goField(forceResolver: true)
   projectId: ID!
   createdAt: DateTime!
-  createdBy: User
+  createdBy: User  @goField(forceResolver: true)
```
Comment (rot1024) [inline server/schemas/asset.graphql:6]: Do not use ` @goField(forceResolver: true)`. Edit `gqlgen.yml` instead

**Review by rot1024 — COMMENTED** [review]
Check codecov warnings. You should add unit tests for uncovered lines

File: server/internal/app/repo.go [hunk 6]
```diff
@@ -35,6 +39,27 @@ func initReposAndGateways(ctx context.Context, conf *Config) (*repo.Container, *
...
+	} else {
+		log.Infof("file: GCS storage is used: %s\n", conf.GCS.BucketName)
+		fileRepo, err = gcs.NewFile(conf.GCS.BucketName, conf.AssetBaseURL, conf.GCS.PublicationCacheControl)
+		if err != nil {
+			if debug {
+				log.Warnf("file: failed to init GCS storage: %s\n", err.Error())
+				err = nil
+			}
+		}
+	}
```
Comment (rot1024) [inline server/internal/app/repo.go:57]:
```suggestion
	var fileRepo gateway.File
	if conf.GCS.BucketName == "" {
		log.Infoln("file: local storage is used")
		datafs := afero.NewBasePathFs(afero.NewOsFs(), "data")
		fileRepo, err = fs.NewFile(datafs, conf.AssetBaseURL)
	} else {
		log.Infof("file: GCS storage is used: %s\n", conf.GCS.BucketName)
		fileRepo, err = gcs.NewFile(conf.GCS.BucketName, conf.AssetBaseURL, conf.GCS.PublicationCacheControl)
		if err != nil {
			log.Fatalf("file: failed to init GCS storage: %s\n", err.Error())
		}
	}
```

- datafs is not used when GCS is used. so it should move inside the more narrow scope
- panic when GCS is failed to be initialized

File: server/internal/app/file.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+				if ct2 != "" {
+					ct = ct2
+				}
+			}
+			return ctx.Stream(http.StatusOK, ct, reader)
+		}
+	}
+
+	ec.GET(
+		"/assets/:filename",
```
Comment (rot1024) [inline server/internal/app/file.go:39]:
```suggestion
		"/assets/:uuid/:filename",
```

- As with GCS, UUID should be required to access assets.
- Support files in directories such as 3D Tiles

e.g.

```
GET /assets/xxxx-xxxx-xxxx-xxxx/hogehoge.zip
GET /assets/xxxx-xxxx-xxxx-xxxx/hogehoge/tileset.json
```

File: server/internal/app/config.go [hunk 8]
```diff
@@ -25,6 +25,8 @@ type Config struct {
 	SMTP         SMTPConfig
 	SendGrid     SendGridConfig
 	SignupSecret string
+	GCS          GCSConfig
+	AssetBaseURL string `default:"http://localhost:8080/assets"`
```
Comment (rot1024) [inline server/internal/app/config.go:29]:
```suggestion
	AssetBaseURL string
```

- In `ReadConfig` func, initialze it with a default value using config.ServerHost
- Rename `ServerHost`field in `Config` struct to `Host`

File: server/internal/adapter/gql/gqlmodel/convert.go [hunk 9]
```diff
@@ -45,3 +51,22 @@ func ToPagination(pagination *Pagination) *usecasex.Pagination {
...
+		ContentType: f.ContentType,
+	}
+}
+
+func ToGCNaming(s string) string {
+	// TODO: we need a better way to implement google storage naming
+	re := regexp.MustCompile(`[^a-zA-Z0-9-.]`)
+	ss := re.ReplaceAllString(s, "-")
+	return strings.ToLower(ss)
+}
```
Comment (rot1024) [inline server/internal/adapter/gql/gqlmodel/convert.go:72]:
```suggestion
```

This function should not be located here as it relates to the internals of GCS. There seems to be an error in the implementation policy when knowledge of GCS appears in this package.

File: server/internal/adapter/gql/gqlmodel/convert_asset.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	if c == nil {
+		return nil
+	}
+
+	var children []*AssetFile
+	for _, a := range c {
+		children = append(children, ToAssetFile(a))
+	}
+	return children
+}
```
Comment (rot1024) [inline server/internal/adapter/gql/gqlmodel/convert_asset.go:79]:
```suggestion
		Children:    lo.Map(a.Children(), func(c *asset.File, _ int) *AssetFile { return ToAssetFile(a) }),
	}
}
```

**Review by rot1024 — CHANGES_REQUESTED** [review]

File: server/internal/app/repo.go (hunk 6 above)
Comment (nourbalaha) [reply server/internal/app/repo.go:57]: Done!

File: server/internal/app/file.go (hunk 7 above)
Comment (nourbalaha) [reply server/internal/app/file.go:39]: I thought that it should be like this:
GET /assets/xx/xxxx-xxxx-xxxx-xxxx/hogehoge/tileset.json

Comment (rot1024) [reply server/internal/app/file.go:39]: OK `/assets/xx/xxx-xxxx-xxxx/hogehoge.zip`

File: server/internal/adapter/gql/gqlmodel/convert_asset.go (hunk 10 above)
Comment (nourbalaha) [reply server/internal/adapter/gql/gqlmodel/convert_asset.go:79]: Done!

File: server/internal/app/file.go (hunk 7 above)
Comment (nourbalaha) [reply server/internal/app/file.go:39]: Done!

File: server/internal/adapter/gql/gqlmodel/convert.go (hunk 9 above)
Comment (nourbalaha) [reply server/internal/adapter/gql/gqlmodel/convert.go:72]: Where should this function be located? can you explain more please?

File: server/internal/app/config.go (hunk 8 above)
Comment (nourbalaha) [reply server/internal/app/config.go]: Done!
Can you check this commit [0e3de14](https://github.com/reearth/reearth-cms/pull/130/commits/0e3de14e2331ae29a1742456afdfe4f0d37bc2c9) please, and let me know if that what you mean.

**Review by rot1024 — APPROVED** [review]

**Pull request merged**